src/storage/src/source/mysql/replication.rs

```diff
@@ -351,12 +352,28 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                             .as_ref()
                             .expect("gtid cap should be set by previous GtidEvent");
 
+                        let end_of_statement = data.flags().contains(RowsEventFlags::STMT_END);
                         events::handle_rows_event(data, &mut repl_context, new_gtid).await?;
 
-                        // Advance the frontier up to the point right before this GTID, since we
-                        // might still see other events that are part of this same GTID, such as
-                        // row events for multiple tables or large row events split into multiple.
-                        if let Err(err) = repl_partitions.update(new_gtid.clone()) {
+                        // If this is the last rows-event for this statement, we know there isn't
+                        // anymore data for this GTID, so we can advance beyond it.
+                        let update_gtid = if end_of_statement {
+                            let mut mut_gtid = next_gtid.take().expect("gtid cap should be set");
+                            // Increment the transaction-id to the next GTID we should see from this source-id
+                            match mut_gtid.timestamp_mut() {
+                                GtidState::Active(time) => {
+                                    *time = time.checked_add(1).unwrap();
+                                }
+                                _ => unreachable!(),
```

Comment:
does mysql always send events in lock step, like this:
```
GtidEvent(10)
RowEvent(...)
RowEvent(...)
GtidEvent(11)
RowEvent(...)
GtidEvent(12)
```

and each gtid event is followed by a _singular transaction_?

Reply:
Yes -- see statement 7 here: https://dev.mysql.com/doc/refman/8.0/en/replication-gtids-lifecycle.html:
> If binary logging is enabled on the replica, the GTID is persisted atomically at commit time by writing it to the binary log at the beginning of the transaction (as a Gtid_log_event). Whenever the binary log is rotated or the server is shut down, the server writes GTIDs for all transactions that were written into the previous binary log file into the mysql.gtid_executed table.

technically there could be other event-types (e.g. QueryEvents which contain DDL):
```
GtidEvent(10)
RowEvent(...)
RowEvent(...)
GtidEvent(11)
QueryEvent(...)
GtidEvent(12)
```

but we should receive the GTID for each event before the corresponding events are received.

Reply:
The reason I ask if, if we got something like this:
```
GtidEvent(10)
RowEvent(...) # for transaction 10, but not told to us
RowEvent(...) # for transaction 10, but not told to us
RowEvent(...) # for transaction 11, but not told to us
GtidEvent(12)
```

then this is incorrect; The docs make it clear that this is not the case! I think linking to those docs to explain this is worth it though!